File: Samples/BasicApplication/BasicApplication/BasicApplication.Script/Administration/Translation/TranslationGrid.cs

namespace BasicApplication.Administration
{
    using jQueryApi;
    using Serenity;
    using System;
    using System.Collections.Generic;
    using System.Runtime.CompilerServices;

    [ColumnsKey("Administration.Translation"), IdProperty("Key")]
    [LocalTextPrefix("Administration.Translation"), Service("Administration/Translation")]
    public class TranslationGrid : EntityGrid<TranslationItem>, IAsyncInit
    {
        private string searchText;
        private LookupEditor sourceLanguage;
        private LookupEditor targetLanguage;
        private string targetLanguageKey;
        private bool hasChanges;

        public TranslationGrid(jQueryObject container)
            : base(container)
        {
            this.element.On("keyup." + this.uniqueName + " change." + this.uniqueName, "input.custom-text", e =>
            {
                var value = J(e.Target).GetValue().TrimToNull();
                if (value == "")
                    value = null;
                this.view.GetItemById(J(e.Target).GetDataValue("key")).CustomText = value;
                hasChanges = true;
            });
        }

        protected override void OnClick(jQueryEvent e, int row, int cell)
        {
            base.OnClick(e, row, cell);

            if (e.IsDefaultPrevented())
                return;

            if (J(e.Target).HasClass("source-text"))
            {
                e.PreventDefault();
                var item = view.Rows[row];

                Action done = delegate
                {
                    item.CustomText = item.SourceText;
                    view.UpdateItem(item.Key, item);
                    hasChanges = true;
                };

                if (item.CustomText.IsTrimmedEmpty() || item.CustomText.TrimToEmpty() == item.SourceText.TrimToEmpty())
                {
                    done();
                    return;
                }

                Q.Confirm(Q.Text("Db.Administration.Translation.OverrideConfirmation"), done);
            }

            if (J(e.Target).HasClass("target-text"))
            {
                e.PreventDefault();
                var item = view.Rows[row];

                Action done = delegate
                {
                    item.CustomText = item.TargetText;
                    view.UpdateItem(item.Key, item);
                    hasChanges = true;
                };

                if (item.CustomText.IsTrimmedEmpty() || item.CustomText.TrimToEmpty() == item.TargetText.TrimToEmpty())
                {
                    done();
                    return;
                }

                Q.Confirm(Q.Text("Db.Administration.Translation.OverrideConfirmation"), done);
            }

        }

        protected override Promise<List<SlickColumn>> GetColumnsAsync()
        {
            var columns = new List<SlickColumn>();
            columns.Add(new SlickColumn { Field = "Key", Width = 300, Sortable = false });
            columns.Add(new SlickColumn
            {
                Field = "SourceText",
                Width = 300,
                Sortable = false,
                Format = ctx =>
                {
                    return jQuery.FromHtml("<a/>")
                        .AddClass("source-text")
                        .Text((string)ctx.Value ?? "")
                        .OuterHtml();
                }
            });
            columns.Add(new SlickColumn
            {
                Field = "CustomText",
                Width = 300,
                Sortable = false,
                Format = ctx =>
                {
                    return jQuery.FromHtml("<input/>")
                        .AddClass("custom-text")
                        .Attribute("value", (string)ctx.Value)
                        .Attribute("type", "text")
                        .Attribute("data-key", (string)ctx.Item.Key)
                        .OuterHtml();
                }
            });
            columns.Add(new SlickColumn
            {
                Field = "TargetText",
                Width = 300,
                Sortable = false,
                Format = ctx =>
                {
                    return jQuery.FromHtml("<a/>")
                        .AddClass("target-text")
                        .Text((string)ctx.Value ?? "")
                        .OuterHtml();
                }
            });

            return Promise.FromValue(columns);
        }

        protected override void CreateToolbarExtensions()
        {
            base.CreateToolbarExtensions();

            sourceLanguage = Widget.Create<LookupEditor>(
                element: e => e.AppendTo(toolbar.Element)
                    .Attribute("placeholder", "--- " + Q.Text("Db.Administration.Translation.SourceLanguage") + " ---"),
                options: new LookupEditorOptions
                {
                    LookupKey = "Administration.Language"
                });

            sourceLanguage.ChangeSelect2(e => 
            {
                if (hasChanges)
                {
                    SaveChanges(targetLanguageKey).Then(Refresh);
                }
                else
                    Refresh();

            });

            targetLanguage = Widget.Create<LookupEditor>(
                element: e => e.AppendTo(toolbar.Element)
                    .Attribute("placeholder", "--- " + Q.Text("Db.Administration.Translation.TargetLanguage") + " ---"),
                options: new LookupEditorOptions
                {
                    LookupKey = "Administration.Language"
                });

            targetLanguage.ChangeSelect2(e => 
            {
                if (hasChanges)
                {
                    SaveChanges(targetLanguageKey).Then(Refresh);
                }
                else
                    Refresh();
            });
        }

        protected virtual Promise SaveChanges(string language)
        {
            var translations = new JsDictionary<string, string>();
            foreach (var item in view.GetItems())
                translations[item.Key] = item.CustomText;

            return Promise.Resolve(TranslationService.Update(new TranslationUpdateRequest
            {
                TargetLanguageID = language,
                Translations = translations
            }, null)).Then(delegate()
            {
                hasChanges = false;
                Q.NotifySuccess("User translations in \"" + language + "\" language are saved to \"user.texts." + language + ".json\" " + 
                    "file under \"~/script/site/texts/user/\"");
            });
        }

        protected override bool OnViewSubmit()
        {
            var request = (TranslationListRequest)view.Params;
            request.SourceLanguageID = sourceLanguage.Value;
            targetLanguageKey = targetLanguage.Value ?? "";
            request.TargetLanguageID = targetLanguageKey;
            hasChanges = false;

            return base.OnViewSubmit();
        } 

        protected override List<ToolButton> GetButtons()
        {
            return new List<ToolButton>
            {
                new ToolButton 
                {
                    Title = "Save Changes",
                    OnClick = e => SaveChanges(targetLanguageKey).Then(Refresh),
                    CssClass = "apply-changes-button"
                }
            };
        }

        protected override void CreateQuickSearchInput()
        {
            GridUtils.AddQuickSearchInputCustom(toolbar.Element, (field, searchText) =>
            {
                this.searchText = searchText;
                this.view.SetItems(this.view.GetItems(), true);
            });
        }

        protected override bool OnViewFilter(TranslationItem item)
        {
            if (!base.OnViewFilter(item))
                return false;

            if (searchText.IsEmptyOrNull())
                return true;

            var searching = Q.Externals.StripDiacritics(searchText).ToLower();
            if (searching.IsEmptyOrNull())
                return true;

            if (Q.Externals.StripDiacritics(item.Key ?? "").ToLower().IndexOf(searching) >= 0)
                return true;

            if (Q.Externals.StripDiacritics(item.SourceText ?? "").ToLower().IndexOf(searching) >= 0)
                return true;

            if (Q.Externals.StripDiacritics(item.TargetText ?? "").ToLower().IndexOf(searching) >= 0)
                return true;

            if (Q.Externals.StripDiacritics(item.CustomText ?? "").ToLower().IndexOf(searching) >= 0)
                return true;

            return false;

        }

        protected override bool UsePager()
        {
            return false;
        }
    }
}

File: serene/src/Serene.Web/Modules/Administration/Translation/TranslationGrid.ts
import { confirmDialog, Decorators, EntityGrid, Fluent, GridUtils, isEmptyOrNull, isTrimmedEmpty, LookupEditor, LookupEditorOptions, notifySuccess, outerHtml, stripDiacritics, ToolButton, trimToEmpty, trimToNull, Widget } from "@serenity-is/corelib";
import { TranslationItem, TranslationTexts } from "@serenity-is/extensions";
import { Column } from "@serenity-is/sleekgrid";
import { TranslationService } from "../../ServerTypes/Administration";

@Decorators.registerClass()
export class TranslationGrid extends EntityGrid<TranslationItem, any> {
    protected getIdProperty() { return "Key"; }
    protected getLocalTextPrefix() { return "Administration.Translation"; }
    protected getService() { return TranslationService.baseUrl; }

    private hasChanges: boolean;
    private searchText: string;
    private sourceLanguage: LookupEditor;
    private targetLanguage: LookupEditor;
    private targetLanguageKey: string;

    constructor(props: any) {
        super(props);

        this.element.on('keyup.' + this.uniqueName + ' change.' + this.uniqueName,
            'input.custom-text', e =>
        {
            var value = trimToNull(Fluent(e.target).val());
            if (value === '') {
                value = null;
            }
            this.view.getItemById(Fluent(e.target).data('key')).CustomText = value;
            this.hasChanges = true;
        });
    }

    protected onClick(e: MouseEvent, row: number, cell: number): any {
        super.onClick(e, row, cell);

        if (e.defaultPrevented || (e as any)?.isDefaultPrevented?.()) {
            return;
        }

        let item = this.itemAt(row);
        let done: () => void;

        if (Fluent(e.target).hasClass('source-text')) {
            e.preventDefault();

            done = () => {
                item.CustomText = item.SourceText;
                this.view.updateItem(item.Key, item);
                this.hasChanges = true;
            };

            if (isTrimmedEmpty(item.CustomText) ||
                (trimToEmpty(item.CustomText) === trimToEmpty(item.SourceText))) {
                done();
                return;
            }

            confirmDialog(TranslationTexts.OverrideConfirmation, done);
            return;
        }

        if (Fluent(e.target).hasClass('target-text')) {
            e.preventDefault();

            done = () => {
                item.CustomText = item.TargetText;
                this.view.updateItem(item.Key, item);
                this.hasChanges = true;
            };

            if (isTrimmedEmpty(item.CustomText) ||
                (trimToEmpty(item.CustomText) === trimToEmpty(item.TargetText))) {
                done();
                return;
            }

            confirmDialog(TranslationTexts.OverrideConfirmation, done);
            return;
        }
    }

    protected getColumns(): Column[] {

        var columns: Column[] = [];
        columns.push({
            field: 'Key',
            name: TranslationTexts.Key,
            width: 300,
            sortable: false
        });

        columns.push({
            field: 'SourceText',
            name: TranslationTexts.SourceText,
            width: 300,
            sortable: false,
            format: ctx => {
                return outerHtml(Fluent('a')
                    .addClass('source-text')
                    .text(ctx.value || ''));
            }
        });

        columns.push({
            field: 'CustomText',
            name: TranslationTexts.CustomText,
            width: 300,
            sortable: false,
            format: ctx => outerHtml(Fluent('input')
                .addClass('custom-text')
                .attr('value', ctx.value)
                .attr('type', 'text')
                .attr('placeholder', ctx.item.TargetText)
                .attr('data-key', ctx.item.Key))
        });

        return columns;
    }

    protected createToolbarExtensions(): void {
        super.createToolbarExtensions();

        let opt: LookupEditorOptions = {
            lookupKey: 'Administration.Language'
        };

        this.sourceLanguage = Widget.create({
            type: LookupEditor,
            element: el => el.appendTo(this.toolbar.element).attr('placeholder', '--- ' +
                TranslationTexts.SourceLanguage + ' ---'),
            options: opt
        });

        this.sourceLanguage.changeSelect2(e => {
            if (this.hasChanges) {
                this.saveChanges(this.targetLanguageKey).then(() => this.refresh());
            }
            else {
                this.refresh();
            }
        });

        this.targetLanguage = Widget.create({
            type: LookupEditor,
            element: el => el.appendTo(this.toolbar.element).attr('placeholder', '--- ' +
                TranslationTexts.TargetLanguage + ' ---'),
            options: opt
        });

        this.targetLanguage.changeSelect2(e => {
            if (this.hasChanges) {
                this.saveChanges(this.targetLanguageKey).then(() => this.refresh());
            }
            else {
                this.refresh();
            }
        });
    }

    protected saveChanges(language: string): PromiseLike<any> {
        var translations: { [key: string]: string } = {};
        for (let item of this.getItems()) {
            translations[item.Key] = item.CustomText;
        }

        return Promise.resolve(TranslationService.Update({
            TargetLanguageID: language,
            Translations: translations
        })).then(() => {
            this.hasChanges = false;
            language = trimToNull(language) || 'invariant';
            notifySuccess('User translations in "' + language +
                '" language are saved to "user.texts.' +
                language + '.json" ' + 'file under "~/App_Data/texts/"', '');
        });
    }

    protected onViewSubmit(): boolean {
        var request = this.view.params;
        request.SourceLanguageID = this.sourceLanguage.value;
        this.targetLanguageKey = this.targetLanguage.value || '';
        request.TargetLanguageID = this.targetLanguageKey;
        this.hasChanges = false;
        return super.onViewSubmit();
    }

    protected getButtons(): ToolButton[] {
        return [{
            title: TranslationTexts.SaveChangesButton,
            onClick: e => this.saveChanges(this.targetLanguageKey).then(() => this.refresh()),
            cssClass: 'apply-changes-button'
        }];
    }

    protected createQuickSearchInput() {
        GridUtils.addQuickSearchInputCustom(this.toolbar.element,
            (field, searchText) => {
                this.searchText = searchText;
                this.view.setItems(this.view.getItems(), true);
            });
    }

    protected onViewFilter(item: TranslationItem) {
        if (!super.onViewFilter(item)) {
            return false;
        }

        if (!this.searchText) {
            return true;
        }

        var sd = stripDiacritics;
        var searching = sd(this.searchText).toLowerCase();

        function match(str: string) {
            if (!str)
                return false;

            return str.toLowerCase().indexOf(searching) >= 0;
        }

        return isEmptyOrNull(searching) || match(item.Key) || match(item.SourceText) ||
            match(item.TargetText) || match(item.CustomText);
    }

    protected usePager() {
        return false;
    }
}